### ISSUE TITLE
Add support for bridged networks in `nixos-container`

### ISSUE THREAD

**Issue by lafka** [NixOS/nixpkgs#5857]
I would like for the containers to add themselves to the bridge on the host to ease network configuration by allowing guests to receive their address from DHCP. `systemd-nspawn` has support for this so it should not be difficult to get a simple hack working. 

My initial plan is to:
 1. add an option `containers.*.bridge :: iface()`
 1. strip away LOCAL_ADDRESS/HOST_ADDRESS from `container-init`
 1. enable DHCP in the container to get address from outside

On the other hand it might be a idea to adapt the container network spec to support more options. I could implement this but first I would like some feedback if this is useful and wanted by the community, pointers on how the actual config should look like.

At the moment I'v looked at adding a `containers.*.networking` option. The spec for that would be roughly:
```
*.networking.type :: Enum "bridge" | "host" | "nat" | "macvlan" | "none" (default: "none")
*.networking.bridge :: iface() when type := "bridge"
*.networking.hostIface :: iface() when type := "host"
*.networking.macvlans :: [iface()] when type := "macvlan"
*.networking.containerAddress :: (optional) ipv4() | ipv6() | dhcp | None (default: None)
*.networking.hostAddress :: (optional) ipv4() | ipv6() | None (default: none) - for providing unique an alias on host interface????
```

This will be implemented with a fallback to the containers root level argument for containerAddress (called localAddress in todays config)  and hostAddress.

I'm uncertain on the exact details of all of the different options but I'l try to summarize (please correct me if I'm wrong)

 * `bridge` - will create a tun/tap pair. One nic will be added to `bridge` on the host and the other will become the nic in the container
 * `host` - the container will bind directly to the host nic specified by `hostIface`
 * `nat` - Creates an isolated network with tun/tap devices and uses iptables to bridge the connection
 * `macvlans` - adds a mac alias to one or more existing interfaces that will be routed to a veth in the container
 * `none` - there will be no networking

What are you guys thinking? Is there a already some plan for this? Is this something for NixOS  or should I aim for a simple hack to fix this?

**Comment by lafka** [comment]
Something like https://github.com/lafka/nixpkgs/commit/0946b3c301ebda8303a408727204adf81fdb188b might do the trick, don't like the imperative way of configuring the container though. This adds `--bridge <iface>` and `--use-dhcp` options to `nixos-configuration create`

Code is only for illustration only, not tested (haven't learned how to patch packages yet)

**Comment by valeriangalliat** [comment]
:+1: for this feature.

**Comment by kampfschlaefer** [comment]
Hi, just a heads up: I implemented ipv6 and host-bridge support for declarative containers in #14018. Maybe thats already what you want?

**Comment by valeriangalliat** [comment]
Thnaks @kampfschlaefer, this should solve the problem I had back in the days (#6208)!

**Comment by kampfschlaefer** [comment]
Hi, with #19128 your container device needs could be fulfilled. With that PR its possible to do all the network definition inside the container, allowing for example vlan on top of the veth that the container is given.

**Comment by lafka** [comment]
@kampfschlaefer it seems you already fixed this back in #14018, just me not paying attention:/ Thanks :)
